File: rhizome/node/seed_node.py
"""
Seed-узел (высокая доступность, долгосрочное хранение)
"""

import asyncio

from rhizome.config import Config
from rhizome.dht.node import Node
from rhizome.node.base_node import BaseNode


class SeedNode(BaseNode):
    """Seed-узел с высокой доступностью и большим объемом хранилища"""

    def __init__(self, config: Config):
        # Принудительно устанавливаем тип seed
        config.node.node_type = "seed"
        super().__init__(config)

    async def start(self):
        """Запуск seed-узла с дополнительными задачами"""
        await super().start()

        # Запуск задач для seed-узла
        asyncio.create_task(self._seed_tasks())

    async def _seed_tasks(self):
        """Фоновые задачи для seed-узла"""
        global_update_interval = self.config.popularity.global_update_interval

        last_global_update = 0

        while self.is_running:
            try:
                current_time = asyncio.get_event_loop().time()

                # Глобальное ранжирование каждые 3 часа
                if current_time - last_global_update >= global_update_interval:
                    await self._update_global_ranking()
                    last_global_update = current_time

                await asyncio.sleep(300)  # Проверяем каждые 5 минут

            except Exception as e:
                self.logger.error("Error in seed tasks", error=str(e), exc_info=True)
                await asyncio.sleep(300)

    async def _update_global_ranking(self):
        """Обновление глобального рейтинга"""
        # Получаем локальные рейтинги
        all_metrics = self.metrics_collector.get_all_metrics()
        if not all_metrics:
            return

        local_ranked = self.popularity_ranker.rank_items(all_metrics, limit=100)

        # Получаем список других seed-узлов
        # TODO: Реализовать получение списка seed-узлов из DHT
        seed_nodes = []  # Пока пустой список

        # Агрегируем глобальный рейтинг
        global_ranking = await self.popularity_exchanger.aggregate_global_ranking(
            local_ranked, seed_nodes
        )

        self.logger.info("Updated global ranking", items=len(global_ranking))

File: src/node/seed_node.rs
use std::ops::Deref;
use std::sync::Arc;
use std::time::Duration;
use tokio::time::sleep;
use tracing::{error, info};

use crate::config::Config;
use crate::node::base_node::{BaseNode, BaseNodePtrs};
use crate::utils::time::get_now_f64;

/// Seed-node for work with popularity
pub struct SeedNode {
    pub base: BaseNode,
}

#[allow(dead_code)]
impl SeedNode {
    pub async fn new(mut config: Config) -> Result<Self, Box<dyn std::error::Error>> {
        config.node.node_type = "seed".to_string();

        let base = BaseNode::new(config).await?;

        Ok(Self { base })
    }

    pub async fn start(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.base.start().await?;

        let base_ptrs = Arc::new(self.base.clone_ptrs());

        tokio::spawn(async move {
            Self::seed_loop(base_ptrs).await;
        });

        info!("Seed-specific tasks started");
        Ok(())
    }

    async fn seed_loop(node: Arc<BaseNodePtrs>) {
        let global_update_interval = node.config.popularity.global_update_interval as f64;
        let mut last_global_update = 0.0;

        while *node.is_running.read().await {
            let current_time = get_now_f64();

            if current_time - last_global_update >= global_update_interval {
                if let Err(e) = Self::update_global_ranking(&node).await {
                    error!(error = %e, "Error updating global ranking in seed task");
                }
                last_global_update = current_time;
            }

            sleep(Duration::from_secs(300)).await;
        }
    }

    async fn update_global_ranking(node: &BaseNodePtrs) -> Result<(), Box<dyn std::error::Error>> {
        let all_metrics = node
            .metrics_collector
            .read()
            .await
            .get_all_metrics()
            .clone();
        if all_metrics.is_empty() {
            return Ok(());
        }

        let local_ranked = node.popularity_ranker.rank_items(&all_metrics, Some(100));

        let mut seed_nodes = Vec::new();
        let all_nodes = node.routing_table.read().await.get_all_nodes();

        // TODO: фильтруем тех, кто похож на seed
        // (Например, по метаданным или отдельному бакету)
        for n in all_nodes {
            seed_nodes.push(n);
        }

        let global_ranking = node
            .popularity_exchanger
            .aggregate_global_ranking(local_ranked, seed_nodes)
            .await;

        info!(
            items = global_ranking.len(),
            "Updated global ranking on seed node"
        );
        Ok(())
    }
}

/// Realization of Deref gives opportunity to call BaseNode methods in SeedNode
/// Example: seed_node.start() <-- seed_node.base.start()
impl Deref for SeedNode {
    type Target = BaseNode;

    fn deref(&self) -> &Self::Target {
        &self.base
    }
}
